File: easy-101-symmetric-tree.cpp
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        return isMirror(root->left, root->right);
    }

private:
    bool isMirror(TreeNode* a, TreeNode* b) {
        if (!a && !b) return true;
        if (!a || !b) return false;
        if (a->val != b->val) return false;
        return isMirror(a->left, b->right) && isMirror(a->right, b->left);
    }
};

class SolutionIter {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        queue<TreeNode*> q;
        q.push(root->left);
        q.push(root->right);

        while (!q.empty()) {
            TreeNode* a = q.front(); q.pop();
            TreeNode* b = q.front(); q.pop();

            if (!a && !b) continue;
            if (!a || !b) return false;
            if (a->val != b->val) return false;

            // mirror pairing
            q.push(a->left);  q.push(b->right);
            q.push(a->right); q.push(b->left);
        }
        return true;
    }
};

File: easy-101-symmetric-tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.isMirror(root.left, root.right)
    
    def isMirror(self, a: TreeNode, b: TreeNode) -> bool:
        if not a and not b:
            return True
        if not a or not b:
            return False
        return (a.val == b.val and
                self.isMirror(a.left, b.right) and
                self.isMirror(a.right, b.left))

from collections import deque

class SolutionIter:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        q = deque([(root.left, root.right)])
        
        while q:
            a, b = q.popleft()
            if not a and not b:
                continue
            if not a or not b:
                return False
            if a.val != b.val:
                return False
            
            # mirror pairing
            q.append((a.left, b.right))
            q.append((a.right, b.left))
        
        return True
